File: src/main/java/LeetCode/RandomPointNonOverlappingRectangles.java
package LeetCode;

import java.util.*;

public class RandomPointNonOverlappingRectangles {
    int numPts;
    List<Integer> rectCumCount;
    int[][] rects;


    public RandomPointNonOverlappingRectangles(int[][] rects) {
        numPts = 0;
        rectCumCount = new ArrayList();
        this.rects = rects;
        for (int[] rect : rects) {
            numPts += (rect[2] - rect[0] + 1) * (rect[3] - rect[1] + 1);
            rectCumCount.add(numPts);
        }
    }

    public int[] pick() {
        int ptIdx = (int)(Math.random() * numPts);
        int l = 0, r = rects.length - 1;
        // Binary search
        while(l < r) {
            int mid = l + (r-l)/2;
            if(rectCumCount.get(mid) <= ptIdx) l = mid + 1;
            else r = mid;
        }
        // l : rectangle index
        int[] rect = rects[l];
        int xPts = rect[2] - rect[0] + 1;
        int yPts = rect[3] - rect[1] + 1;
        int ptsInRect = xPts * yPts;
        int ptStart = rectCumCount.get(l) - ptsInRect;
        int offset = ptIdx - ptStart;
        return new int[] {rect[0] + offset%xPts, rect[1] + offset/xPts};
    }
}

File: python/leetcode/random_point_non-overlapping_rectangles.py
class Solution(object):

    def __init__(self, rects):
        """
        :type rects: List[List[int]]
        """
        self.numPts = 0;
        self.rectCumCount = []
        self.rects = rects;
        for rect in rects:
            self.numPts += (rect[2] - rect[0] + 1) * (rect[3] - rect[1] + 1)
            self.rectCumCount.append(self.numPts)

    def pick(self):
        """
        :rtype: List[int]
        """
        ptIdx = random.randint(0, self.numPts - 1)
        l = 0; r = len(self.rects) - 1
        # Binary search
        while l < r:
            mid = l + (r-l)/2
            if self.rectCumCount[mid] <= ptIdx: l = mid + 1
            else: r = mid
        # l : rectangle index
        rect = self.rects[l]
        xPts = rect[2] - rect[0] + 1
        yPts = rect[3] - rect[1] + 1
        ptsInRect = xPts * yPts
        ptStart = self.rectCumCount[l] - ptsInRect
        offset = ptIdx - ptStart
        return [rect[0] + offset%xPts, rect[1] + offset/xPts]



# Your Solution object will be instantiated and called as such:
# obj = Solution(rects)
# param_1 = obj.pick()